File: leetcode_problems/python/lc680_valid_palindrome_2.py
class Solution:
    def help(self, s: str, p1: int, p2: int) -> bool:
        while p1 < p2:
            if s[p1] != s[p2]:
                return False
            p1 += 1
            p2 -= 1
        return True

    def validPalindrome(self, s: str) -> bool:
        p1 = 0
        p2 = len(s) - 1

        while p1 < p2:
            char1 = s[p1]
            char2 = s[p2]

            if char1 == char2:
                p1 += 1
                p2 -= 1

            else:
                return self.help(s, p1 + 1, p2) or self.help(s, p1, p2 - 1)

        return True


if __name__ == "__main__":
    s = "abca"
    s = "abc"
    s = "aguokepatgbnvfqmgmlcupuufxoohdfpgjdmysgvhmvffcnqxjjxqncffvmhvgsymdjgpfdhooxfuupuculmgmqfvnbgtapekouga"

    print(Solution().validPalindrome(s))

File: leetcode_problems/cpp/lc680_valid_palindrome_2.cpp
#include <string>
#include <iostream>
using std::string;

class Solution {
public:
    bool isPalindrome(const string& s, int p1, int p2) {
        while (p1 < p2) {
            if (s[p1] != s[p2]) {
                return false;
            }
            p1++;
            p2--;
        }
        return true;
    }

    bool validPalindrome(string s) {
        int p1 = 0;
        int p2 = s.size() - 1;
        while (p1 < p2) {
            if (s[p1] != s[p2]) {
                return isPalindrome(s, p1 + 1, p2) || isPalindrome(s, p1, p2 - 1);
            }
            p1++;
            p2--;
        }
        return true;
    }
};

int main() {
    string s = "abca";
    Solution sol;
    bool result = sol.validPalindrome(s);
    std::cout << (result ? "true" : "false") << std::endl;
    return 0;
}